core/src/avm2/globals/flash/utils/bytearray.rs

```diff
@@ -0,0 +1,946 @@
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        if let Ok(num) = bytearray.read_double() {
+            return Ok(Value::Number(num));
+        }
+    }
+    Ok(Value::Undefined)
+}
+
+pub fn read_float<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        if let Ok(num) = bytearray.read_float() {
+            return Ok(Value::Number(num as f64));
+        }
+    }
+    Ok(Value::Undefined)
+}
+
+pub fn read_int<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        if let Ok(num) = bytearray.read_int() {
+            return Ok(Value::Integer(num));
+        }
+    }
+    Ok(Value::Undefined)
+}
+
+pub fn read_unsigned_int<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        if let Ok(num) = bytearray.read_unsigned_int() {
+            return Ok(Value::Unsigned(num));
+        }
+    }
+    Ok(Value::Undefined)
+}
+
+pub fn read_boolean<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        if let Ok(num) = bytearray.read_boolean() {
+            return Ok(Value::Bool(num));
+        }
+    }
+    Ok(Value::Undefined)
+}
+
+pub fn read_byte<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    _args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        if let Ok(num) = bytearray.read_byte() {
+            return Ok(Value::Integer(num as i32));
+        }
+    }
+    Ok(Value::Undefined)
+}
+
+pub fn read_utf_bytes<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(mut bytearray) = this
+        .unwrap()
+        .as_bytearray_mut(activation.context.gc_context)
+    {
+        if let Some(Value::Integer(len)) = args.get(0) {
+            if *len < 0 {
+                log::error!("ByteArray: Did not get proper length");
+                return Ok(Value::Undefined);
+            }
+            if let Ok(bytes) = bytearray.read_exactly(*len as usize) {
```

Comment:
`read_exactly` yields an `io::Result`; the error should be converted into an AVM2 error and then "thrown" (tried). AS3 spec specifically calls out `EOFError`; but since we don't have typed errors yet (or any sort of exception handling) I think it's safe to just `.map_err(|_| "EOFError: Not enough data in byte array!")` and be done with it for now.